src/diff/changes/categorizedChange.ts

```diff
@@ -0,0 +1,30 @@
+/*
+ * Copyright (c) 2020, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { RuleCategory } from "../ruleSet";
+
+/**
+ * Holds node change that is categorized by rules
+ */
+export class CategorizedChange {
+  /**
+   * @param ruleName - Name of the rule that is applied/passed
+   * @param ruleEvent - Event type defined in the rule
+   * @param category - Category defined in the rule. Example: Breaking, Non-Breaking, Ignore
+   * @param ignored - Flag indicating if a change is ignored by the rule
+   * @param change - actual value that is changed - [oldValue, newValue]; optional if the rule is ignored
+   */
+  constructor(
+    public ruleName: string,
+    public ruleEvent: string,
+    public category: RuleCategory,
+    public change?: [unknown, unknown]
+  ) {
```

Comment:
```suggestion
export class CategorizedChange<C extends RuleCategory> {
  /**
   * @param ruleName - Name of the rule that is applied/passed
   * @param ruleEvent - Event type defined in the rule
   * @param category - Category defined in the rule. Example: Breaking, Non-Breaking, Ignore
   * @param ignored - Flag indicating if a change is ignored by the rule
   * @param change - actual value that is changed - [oldValue, newValue]; optional if the rule is ignored
   */
  constructor(
    public ruleName: string,
    public ruleEvent: string,
    public category: RuleCategory,
    public change?: C extends RuleCategory.IGNORED ? undefined : [unknown, unknown]
  ) {
```
This will ensure that TypeScript reports an error if a change is passed along with category `IGNORED`.

Reply:
change is allowed with ignored, its just optional because its hard to determine the change